### PR TITLE
Env Flag Filtering

### PR DESCRIPTION
**Pull request by maxcoulombe**
### Summary
+ added filtering for env flags when generating command parsing warnings
- removed some duplication and created constants so new flags are properly handled automatically if added to the defined collection

### Problem
Env flags passed at the end of a command generates undesired warnings, e.g.:
```
vault read sys/auth/approle/tune -format=json
Flags must be provided before positional arguments. The following arguments will not be parsed as flags: [-format=json]
(...)
```

### Context
Some flags are not tied to a command but to the general execution environment and are valid when passed at the very end. For example: `format=json`. Warnings introduced during [this pull request](https://github.com/hashicorp/vault/pull/16441) could create false positives when environment flags are used appropriately as it treated any trailing flags as a potential user mistake.

With this fix env flags are acceptable both as flags `options` and `args`
`vault <command> [options] [path] [args]`

### Example
```
$ vault read sys/auth/approle/tune --format=json
{
  "request_id": "33cafd05-197f-3fcb-53ba-bc1bf3689f23",
  "lease_id": "",
  "lease_duration": 0,
  "renewable": false,
  "data": {
    "default_lease_ttl": 2764800,
    "description": "",
    "force_no_cache": false,
    "max_lease_ttl": 2764800,
    "token_type": "default-service"
  },
  "warnings": [
    "Endpoint ignored these unrecognized parameters: [--format]"
  ]
}

$ vault read --format=json sys/auth/approle/tune
{
  "request_id": "2386f2f2-8c5c-915b-ffe3-7df7941cc088",
  "lease_id": "",
  "lease_duration": 0,
  "renewable": false,
  "data": {
    "default_lease_ttl": 2764800,
    "description": "",
    "force_no_cache": false,
    "max_lease_ttl": 2764800,
    "token_type": "default-service"
  },
  "warnings": null
}
```

### CODE REVIEW

File: command/base_helpers.go [hunk 1]
```diff
@@ -296,13 +296,26 @@ func parseFlagFile(raw string) (string, error) {
 func generateFlagWarnings(args []string) string {
 	var trailingFlags []string
 	for _, arg := range args {
-		if strings.HasPrefix(arg, "-") {
-			trailingFlags = append(trailingFlags, arg)
+		if !strings.HasPrefix(arg, "-") {
+			continue
 		}
+
+		isEnvFlag := false
```
Comment (tomhjp) [inline command/base_helpers.go:303]: I can see what it's getting at, but "environmental flag" feels a bit unclear to me, especially in the context of flag/option parsing where it could be confused to mean "flags that use environment variables" in some way. Would global be a better adjective to use? i.e. they are flags that can appear anywhere in the command line for almost any command, so they're global flags.

File: command/base_helpers_test.go [hunk 2]
```diff
@@ -242,6 +242,22 @@ func TestArgWarnings(t *testing.T) {
 			[]string{"-a", "b"},
 			"-a",
 		},
+		{
+			[]string{envFlagDetailed},
+			"",
+		},
+		{
+			[]string{"-" + envFlagOutputCurlString + "=true"},
```
Comment (tomhjp) [inline command/base_helpers_test.go:250]: Maybe worth an additional test case with the `--` form as well?

File: command/main.go [hunk 3]
```diff
@@ -24,6 +24,17 @@ type VaultUI struct {
...
 
+const (
+	envFlagOutputCurlString = "output-curl-string"
+	envFlagOutputPolicy     = "output-policy"
+	envFlagFormat           = "format"
+	envFlagDetailed         = "detailed"
+)
+
+var envFlags = []string{
+	envFlagOutputCurlString, envFlagOutputPolicy, envFlagFormat, envFlagDetailed,
```
Comment (tomhjp) [inline command/main.go:35]: As above, `globalFlagXXX` would be slightly clearer to me, but feel free to disagree

**Review by tomhjp — APPROVED** [review]
LGTM, just naming + nits

**Comment by tomhjp** [comment]
The warnings within the JSON response payload are due to this PR: https://github.com/hashicorp/vault/pull/14962. Is that what you're referring to as a breaking change, or the CLI warning "`The following arguments will not be parsed as flags: [--foo=bar]`"?

It does seem odd though that `--format` is included in the JSON payload warnings.

**Comment by fairclothjm** [comment]
@tomhjp The breaking change I was referring to is the inclusion of `Command flags must be provided before positional arguments. The following arguments will not be parsed as flags: [--foo=bar]` before the returned json response. Users that were parsing json with tools like `jq` will be broken. I consider this to be a quite common use case so I am quite sensitive to it. Maybe this is ok though and we move forward but make sure to document the change?

As for the warnings, yes at first I was a bit confused. But maybe it makes sense if "endpoint" only refers to the `/tune` portion? But if by endpoint we mean Vault then I am not sure it makes sense.

File: command/base_helpers.go (hunk 1 above)
Comment (maxcoulombe) [reply command/base_helpers.go:303]: I agree globalFlag to denote flags that apply to any command might be more clear. I'll update the naming.

**Comment by maxcoulombe** [comment]
I agree if the format is explicitly specified as json the output should be json-valid no matter the circumstances such as potential warnings.

I added a check on the desired format before printing the warnings, let me know if you think it's a proper way of hangling this.

Sample results:
```
max@max-Precision-7560:~/projects/vault$ vlto kv put -mount=secret foo bar=baz -asd=qwe
Command flags must be provided before positional arguments. The following arguments will not be parsed as flags: [-asd=qwe]
= Secret Path =
secret/data/foo

======= Metadata =======
Key                Value
---                -----
created_time       2022-08-17T18:33:45.196800638Z
custom_metadata    <nil>
deletion_time      n/a
destroyed          false
version            3
max@max-Precision-7560:~/projects/vault$ vlto kv put -mount=secret foo bar=baz -asd=qwe --format=json
{
  "request_id": "6a317946-18b0-618f-6431-67ba5bfcd351",
  "lease_id": "",
  "lease_duration": 0,
  "renewable": false,
  "data": {
    "created_time": "2022-08-17T18:33:46.650979208Z",
    "custom_metadata": null,
    "deletion_time": "",
    "destroyed": false,
    "version": 4
  },
  "warnings": null
}
max@max-Precision-7560:~/projects/vault
```

File: command/base.go [hunk 4]
```diff
@@ -589,7 +589,7 @@ func (f *FlagSets) Parse(args []string) error {
 	err := f.mainSet.Parse(args)
 
 	warnings := generateFlagWarnings(f.Args())
-	if warnings != "" {
+	if warnings != "" && Format(f.ui) != "json" {
```
Comment (tomhjp) [inline command/base.go:592]: Should we also exclude yaml? Or maybe it's simpler to have an allow list, probably table and pretty? Full list of supported formatters here: https://github.com/hashicorp/vault/blob/7d08dcc21a1d4cd6f7974e3e60408b1e4d43fec9/command/format.go#L67-L73

File: command/main.go [hunk 5]
```diff
@@ -24,6 +24,17 @@ type VaultUI struct {
...
 }
 
+const (
+	globalFlagOutputCurlString = "output-curl-string"
+	globalFlagOutputPolicy     = "output-policy"
+	globalFlagFormat           = "format"
+	globalFlagDetailed         = "detailed"
+)
+
+var envFlags = []string{
```
Comment (tomhjp) [inline command/main.go:34]:
```suggestion
var globalFlags = []string{
```

File: command/main.go [hunk 6]
```diff
@@ -115,6 +116,20 @@ func setupEnv(args []string) (retArgs []string, format string, detailed bool, ou
 	return args, format, detailed, outputCurlString, outputPolicy
 }
 
+func isEnvFlag(arg string, flag string) bool {
```
Comment (tomhjp) [inline command/main.go:119]:
```suggestion
func isGlobalFlag(arg string, flag string) bool {
```

File: command/main.go [hunk 7]
```diff
@@ -115,6 +116,20 @@ func setupEnv(args []string) (retArgs []string, format string, detailed bool, ou
...
+	return arg == "-"+flag || arg == "--"+flag
+}
+
+func isEnvFlagWithValue(arg string, flag string) bool {
```
Comment (tomhjp) [inline command/main.go:123]:
```suggestion
func isGlobalFlagWithValue(arg string, flag string) bool {
```

File: command/main.go [hunk 8]
```diff
@@ -115,6 +116,20 @@ func setupEnv(args []string) (retArgs []string, format string, detailed bool, ou
...
+	return strings.HasPrefix(arg, "--"+flag+"=") || strings.HasPrefix(arg, "-"+flag+"=")
+}
+
+func getEnvFlagValue(arg string) string {
```
Comment (tomhjp) [inline command/main.go:127]:
```suggestion
func getGlobalFlagValue(arg string) string {
```

**Review by tomhjp — COMMENTED** [review]
Good point about the JSON output 👍 agreed on the approach here

File: command/base.go (hunk 4 above)
Comment (fairclothjm) [reply command/base.go:592]: +1

Also, would it be possible to capture these scenarios in some tests? Unit or integration?

Comment (maxcoulombe) [reply command/base.go]: Do you have a link or example of integration tests that validate the format of some outputs?

Comment (fairclothjm) [reply command/base.go]: Looks like pretty is for custom formatters. Currently it is only implemented for the [raft autopilot state API](https://github.com/hashicorp/vault/blob/195e025cb2646209ccd49caef395690cbcdb7bd3/command/format.go#L151)

Comment (fairclothjm) [reply command/base.go]: > Do you have a link or example of integration tests that validate the format of some outputs?

I realize many of these cases are more like unplanned "features" that just happened as vault grew. And are probably very unlikely edge cases. So it may not be worthwhile.

**Review by fairclothjm — APPROVED** [review]
